packages/@aws-cdk/aws-synthetics/lib/canary.ts

```diff
@@ -0,0 +1,255 @@
+import * as iam from '@aws-cdk/aws-iam';
+import * as s3 from '@aws-cdk/aws-s3';
+import * as cdk from '@aws-cdk/core';
+import { CanaryBase } from './canary-base';
+import { CfnCanary } from './synthetics.generated';
+
+/**
+ * Optional properties for a canary
+ */
+export interface CanaryOptions extends cdk.ResourceProps {
+  /**
+   * The s3 location that stores the data of the canary runs.
+   *
+   * @default - A new s3 bucket will be created.
+   */
+  readonly artifactBucket?: s3.IBucket;
+
+  /**
+   * Canary execution role.
+   *
+   * This is the role that will be assumed by the canary upon execution.
+   * It controls the permissions that the canary will have. The Role must
+   * be assumable by the 'lambda.amazonaws.com' service principal.
+   *
+   * The default Role automatically has permissions granted for Canary execution.
+   * If you provide a Role, you must add the relevant policies yourself.
+   *
+   * The relevant policies are "s3:PutObject", "s3:GetBucketLocation", "s3:ListAllMyBuckets",
+   * "cloudwatch:PutMetricData", "logs:CreateLogGroup", "logs:CreateLogStream", "logs:PutLogEvents".
+   *
+   * @default - A unique role will be generated for this canary.
+   * Both supplied and generated roles can always be changed by calling 'addToRolePolicy'.
+   */
+  readonly role?: iam.IRole;
+
+  /**
+   * How many seconds the canary should run before timing out.
+   *
+   * @default - the smaller value between frequency and 900 seconds
+   */
+  readonly timeout?: cdk.Duration;
+
+  /**
+   * The amount of memory, in MB, that is allocated to your Canary. Must be a multiple of 64 and inside the range 960 to 3008.
+   *
+   * @default - 960
+   */
+  readonly memorySize?: number;
+
+  /**
+   * How long the canary will be in a 'RUNNING' state. For example, if you set `timeToLive` to be 1 hour and `frequency` to be 10 minutes,
+   * your canary will run at 10 minute intervals for an hour, for a total of 6 times.
+   *
+   * The default of 0 seconds means that the canary will continue to make runs at the specified frequency until you stop it.
+   *
+   * @default Duration.seconds(0)
+   */
+  readonly timeToLive?: cdk.Duration;
+
+  /**
+   * How often the canary will run during its lifetime. The syntax for expression is 'rate(number unit)'
+   * where unit can be 'minute', 'minutes', or 'hour'. You can specify a frequency between 'rate(1 minute)'
+   * and 'rate(1 hour)'.
+   *
+   * The special expression 'rate(0 minute)' specifies that the canary will run only once when it is started.
+   *
+   * @default 'rate(5 minutes)'
+   */
+  readonly frequency?: cdk.Duration;
```

Comment:
Hmmm... there's a stark contrast between your `frequency` type and the documentation that goes with it. The fact CFN gets an expression in the form of `rate(...)` leads me to believe maybe there will be other kinds of triggers in the future... Maybe it should be somewhat differently?

Look for other places where "rate()" is used.

Reply:
As far as I can tell, there are many places that utilize a `Schedule` class and there is a GitHub issue to make a `Schedule` class in cdk core: #8979. They all have some sort of `rate()` method as well as `cron()` and `expression()`. 

For this use case, however, it seems like the Canary `Schedule` property is not as complicated and does not currently offer anything other than a `rate`. There are a few ways to model the API here. 

Option 1: Keep `frequency` as is - it describes what the CFN `Schedule` represents and we do not need to make it more complicated.

```ts
frequency: cdk.Duration,
```

Option 2: Develop a `Schedule` class in core and use that as the unit for a `schedule` property in the canary. It is a bit more verbose but will allow for other types of schedules in the future.

```ts
schedule: cdk.Schedule.rate(cdk.Duration),
```

A quick sketch of this class could be:

```ts
class Schedule {
  // instantiation
  public static rate(duration: Duration){}

  public static at(moment: Date){}

  public static cron(options: CronOptions){}

  // conversion
  public static toRate(){}
  
  // more conversion methods as needed
}
```

I'm leaning towards Option 2 as I feel it will be nice to move a commonly used pattern to the cdk core. I'm interested in soliciting other opinions here though.

Reply:
I agree we need a `Schedule`, lets implement one in `core` and use it in synthetics and all others locations mentioned in  https://github.com/aws/aws-cdk/issues/8979. This should be a separate PR